src/array.rs

```diff
@@ -183,6 +187,150 @@ impl<'a, A: Clone> TreeLike for Partition<'a, A> {
     }
 }
 
+/// Convert a tree into a binary tree.
+/// Each node has fanout at most two.
+///
+/// Nodes with at most two children (non-arrays) are preserved in the binary tree.
+/// We call these nodes _normal nodes_.
+///
+/// Nodes with more than two children (arrays) are converted into balanced binary trees.
+///
+/// ## Example
+///
+/// Take a root node with children `a`, `b`, `c`.
+///
+/// ```text
+///    .
+///  / | \
+/// a  b  c
+/// ```
+///
+/// The 3-ary root is converted into two 2-ary nodes. The normal nodes `a`, `b`, `c` stay preserved.
+///
+/// ```text
+///     .
+///    / \
+///   .   c
+///  / \
+/// a   b
+/// ```
+#[derive(Debug, Clone, Eq, PartialEq, Hash)]
+pub struct BinaryTree<T>(BinaryTreeInner<T>);
+
+impl<T> BinaryTree<T> {
+    /// Access the content of a normal node.
+    pub fn as_normal(&self) -> Option<&T> {
+        match &self.0 {
+            BinaryTreeInner::Normal(tree) => Some(tree),
+            _ => None,
+        }
+    }
+}
+
+impl<T: TreeLike> BinaryTree<T> {
+    /// Convert a tree into a binary tree.
+    ///
+    /// ## Panics
+    ///
+    /// `<T as TreeLike>::as_node` returns `Tree::Nary` with an empty array.
```

Comment:
In e2e0fd2a0f674e396da6c0e6f5045d86bc362a76:

How can this panic ever happen? It seems like your constructors all prevent this from happening.

Reply:
Same with a few other `# Panics` sections below.

Reply:
Unfortunately this can happen. `BinaryTree` ultimately wraps an existing `T: TreeLike`. When `tree.as_node()` returns `Tree::Nary(array)` with `array.len() = 0` then there will be a panic in `BinaryTreeInner::array`.

`Tree::Nary` is supposed to be used for nodes with more than two children, but there is nothing that stops implementors from ignoring that convention.

Reply:
Gotcha, thanks!